extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -61,23 +70,63 @@ private static QuarkusSecurityIdentity augmentIdentity(SecurityIdentity security
                     public Uni<Boolean> apply(Permission permission) {
                         return securityIdentity.checkPermission(permission);
                     }
-                })
-                .build();
+                }).build();
     }
 
+    /**
+     * A bounded map that limits the number of refresh tokens we can process concurrently
+     */
+    private final Map<String, TokenEntry> refreshingTokens = Collections
+            .synchronizedMap(new LinkedHashMap<String, TokenEntry>(16, 0.75f, true) {
+                @Override
+                protected boolean removeEldestEntry(Map.Entry eldest) {
+                    // we could make the max number of entries configurable
+                    return size() > 30;
+                }
+            });
+
     public Uni<SecurityIdentity> authenticate(RoutingContext context,
             IdentityProviderManager identityProviderManager,
             DefaultTenantConfigResolver resolver) {
         Cookie sessionCookie = context.request().getCookie(SESSION_COOKIE_NAME);
+        TenantConfigContext configContext = resolver.resolve(context, true);
 
         // if session already established, try to re-authenticate
         if (sessionCookie != null) {
             String[] tokens = sessionCookie.getValue().split(COOKIE_DELIM);
+            String idToken = tokens[0];
+            String accessToken = tokens[1];
+            String refreshToken = tokens[2];
+
             return authenticate(identityProviderManager, new IdTokenCredential(tokens[0], context))
                     .map(new Function<SecurityIdentity, SecurityIdentity>() {
                         @Override
-                        public SecurityIdentity apply(SecurityIdentity securityIdentity) {
-                            return augmentIdentity(securityIdentity, tokens[1], tokens[2], context);
+                        public SecurityIdentity apply(SecurityIdentity identity) {
+                            if (isLogout(context, configContext)) {
+                                // force logout, the challenge will be built accordingly and redirect to the logout endpoint at the OP
+                                throw new AuthenticationFailedException();
+                            }
+
+                            return augmentIdentity(identity, accessToken, refreshToken, context);
+                        }
+                    }).on().failure().recoverWithItem(new Function<Throwable, SecurityIdentity>() {
+                        @Override
+                        public SecurityIdentity apply(Throwable throwable) {
+                            SecurityIdentity identity;
+
+                            // if authentication failed, the reason could be the token no longer being valid, so we try a 
+                            // silent refresh if required
+                            if (isLogout(context, configContext) ||
```

Comment:
@pedroigor That seem different to what was done about the logout earlier where a challenge was attempted with a valid identity only. IMHO the refresh can not be attempted if the token has expired, which can happen on any other request, so IMHO your original flow should be restored (the logout request initiates a challenge flow only if the identity is available) and again, please just check if `"Token has expired".equals(throwable.getMessage())` :-)

Reply:
I was thinking more about it, so yes, IMHO, the token refresh should be treated on its own, since the logout request, like any other request to the secured endpoint, should be authenticated first. Next, if the token is valid, and it is a logout, then we initiate a logout flow. If the token verification fails due to the token being expired then we try to recover and refresh ID token as per your idea - but it will work for any other secured request :-)

Reply:
@pedroigor Hey Pedro, the other thing I forgot. Awhile back I introduced `AuthenticationRedirectException` to handle the local redirect (as part of the authentication request when the user is coming back from the OIDC and the original URI is restored) as I felt using the challenge method for that (ex, throw `AuthenticationFailedException` and then check again in the challenge method if the redirect is needed) was not ideal (in fact the redirect to drop the query parameters falls into the same category).
So IMHO it would a bit simpler, if in the authentication method, once the token has been verified and isLogout is confirmed then you create there a redirect URI and throw `AuthenticationRedirectException` - no challenge method will even be invoked - which is also good - as it is not really a challenge that we do here but redirect to OIDC to logout...

Reply:
OK. We can use that exception to redirect.

Reply:
@sberyozkin Just noticed that using the `AuthenticationRedirectException` would be against one of your suggestions to avoid multiple calls to `removeCookie`.

I think we should leave the code as is and have the invalidation code in a single place, as you previously suggested.

Reply:
@pedroigor 

> Just noticed that using the AuthenticationRedirectException would be against one of your suggestions to avoid multiple calls to removeCookie.

What do you mean ?

Reply:
I see. I thought that you would like to have the local invalidation from a single point in the code. That said, using the redirect exception is going to spread local invalidation to other parts of the code.

Reply:
@pedroigor 
> using the redirect exception is going to spread local invalidation to other parts of the code
You mean in the same class ? Is it a blocker to having a non-challenge redirect be handled consistently and without the double logout check ?

Reply:
Yes. For the sake of consistency, it makes more sense to me to have removed the cookie always from a single place. In this case, whenever a challenge is made.

But I don't mind using the redirect exception and calling `removeCookie` from different places.